Toc keyboard navigation

astephanus opened the pull request:
### Description
Adds keyboard navigation to the left navigation TOC using tab, space, enter / return, and arrow keys in accordance with the W3C ARIA Authroing Practices Guide definition of the Tree View Pattern 


Adds missing role,  aria-expanded, aria-current, aria-owned, corresponding ID attributes to the left navigation TOC tree in accordance with the W3C specification and working example.

Includes style fixes for focus rectangles.

Reference documentation
https://www.w3.org/WAI/ARIA/apg/patterns/treeview/
https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-navigation/

### Issues Resolved
https://github.com/opensearch-project/documentation-website/issues/518


### Checklist
- [ ] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ashwin-pc commented on _layouts/default.html:
```diff
@@ -66,34 +66,54 @@
             | size %}
         {% if pages_top_size > 0 %}
-          {% include nav.html pages=site.html_pages key=nil %}
+          {% include nav.html pages=site.html_pages key=nil tab_index="1" expanded=true %}
```
I dont think this is a good idea. It looks like you made this change to get around the keyboard trap caused by the search bar, but this is not a good idea since this makes navigating to the rest of the app harder.

ashwin-pc reviewed (commented):
A few high level comments here:
1. Dont change the tab order unless necessary. The current default of starting with the top nav is useful because i can navigate away from the page easily and go to the about and other pages easily. This change the order making getting to those pages much harder.
2. When you enable keyboard navigation for the tree view, also make sure to remove tab navigation through the tree. This makes navigating to the rest of the page much easier.
3. The keyboard navigation is good but can be improved. For example, in an unexpanded state, the down arrow key should only go to the row that it has to expand, and not to the arrow itself. And the right and left arrow keys should be used to expand and collapse a tree view node. See the W3C example on how this navigation should work: https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-navigation/
4. There seems to be a keyboard trap in the search bar which makes getting to the navigation drawer impossible. Lets fix that i this change since this change wont really improve keyboard accessibility without fixing that too.

astephanus commented:
@ashwin-pc Following up on the feeback I have removed the focus trap on the search field, as well as the explicit values for tabindex allowing for all focusble elements in the page to be reachable by TAB in their natural DOM order. I have also implemented the arrow key navigation according to the referenceced Tree View Navigation pattern.  I also removed my initial feature of defaulting keyboard focus to the aria-current="page" navigation item on page load to allow for natural TAB flow. During a demonstration and discussion with the team this morning it was decided upon maintaining the TAB navigation through the navigation tree in addition to the arrow navigation to maintain that all focusable TAB reachable elements in the page are reachable by the TAB key. TAB and SHIFT+TAB move down and up through the list respectively according to that which is or is not expanded. SPACE will expand / collapse an expandable tree when focused and ENTER will navigate to that item as well.

hdhalter commented:
@ashwin-pc - Which browser are you using to test? We discovered that Chrome and Firefox behave very differently. Aaron's latest fixes may work in Chrome, but they do not work in Firefox. He'll go back and make sure they work in all browsers we support. 

Supported browsers are documented here: http://localhost:4000/docs/latest/install-and-configure/install-dashboards/index/#browser-compatibility. @kgcreative - please confirm.

astephanus commented:
@hdhalter @ashwin-pc Curiously enough, there are some configuration settings that are required. My Chrome instance had it set by default. I do not know if this is a Mac only issue. I will look for a Windows testing VM if those are still available on MSDN. For reference the following document on the a11y Project's site discusses this matter: https://www.a11yproject.com/posts/macos-browser-keyboard-navigation/

ashwin-pc commented:
> During a demonstration and discussion with the team this morning it was decided upon maintaining the TAB navigation through the navigation tree in addition to the arrow navigation to maintain that all focusable TAB reachable elements in the page are reachable by the TAB key

What was the reason for keeping this? It makes keyboard navigation so much harder for a disabled user. Especially if the tree is large with no added benefit.

The example used in the issue description is for top navigation where both tab and keyboard navigation make sense: 
> Adds keyboard navigation to the left navigation TOC using tab, space, enter / return, and arrow keys in accordance with the W3C ARIA Authroing Practices Guide definition of the "Disclosure Pattern": https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-navigation-hybrid/

But the docsite and project website do not use a top nav and instead use tree navigation where the W3C recommends eliminating the tab navigation. This is for good reason since tree view lists can be much longer to navigate than topnav elements, especially once expanded. By removing tabbing we make navigating from the tree to the content a lot easier.

Ref tree view W3C recommendation: https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-navigation/

> @ashwin-pc - Which browser are you using to test? We discovered that Chrome and Firefox behave very differently. Aaron's latest fixes may work in Chrome, but they do not work in Firefox. He'll go back and make sure they work in all browsers we support.

@hdhalter my testing was done in a Chromium based browser on MacOS. I do understand that there are differences between browsers and operating systems which is why i used https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-navigation/ as a reference on the same browser and operating system and it works as expected there.

astephanus commented:
> > During a demonstration and discussion with the team this morning it was decided upon maintaining the TAB navigation through the navigation tree in addition to the arrow navigation to maintain that all focusable TAB reachable elements in the page are reachable by the TAB key
> 
> What was the reason for keeping this? It makes keyboard navigation so much harder for a disabled user. Especially if the tree is large with no added benefit.

The choice was driven by the idea that every interactive element on the page be reachable by the TAB key. There is an inherit ambiguity to the notion of `element`. The navigation tree itself could be conceived of as a single element. Of course, the usability challenge of needing to TAB through dozens of links to get to the actual content is obvious.

I followed up with the keyboard navigation implementation using the W3C example. The active navigation item is reachable by TAB given `tabindex="0"` falling naturally in DOM order. The remaining items are removed from tab order using `tabindex="-1"`.  What's a little weird is that the page header with "Documentation" H1 and the search field actually comes after the left panel even though it appears above the left panel unituitively. There are other template  and style peculiarities which seem related to the Just the Docs theme, and the aim is to make as few changes, and keep things as simple as possible with this PR while satisfying the defect and the discussions about keyboard navigation.

I have tested successfully on Mac with Chrome, Firefox, and Safari.

ashwin-pc reviewed (approved):
@astephanus This is much better thank you. Behavior you are seeing with the unintuitive tab order is because the H1 and the search bar appear later in the HTML markup than the side panel elements. You can change the other that they are added to the HTML to fix this but i dont think thats needed as a part of this PR.

Aside from my one comment about the custom style changes there is one other thing i noticed with the top nav bar elements not being tabbable. Only 2 of the 4 tab options are tabbable but thats something that already exists today but we should fix before we can claim to have resolved accessibility issues

astephanus commented:
> @astephanus This is much better thank you. Behavior you are seeing with the unintuitive tab order is because the H1 and the search bar appear later in the HTML markup than the side panel elements. You can change the other that they are added to the HTML to fix this but i dont think thats needed as a part of this PR.

Thank you. I figured changing the DOM order of those elements might be beyond the scope of this PR. I think it makes sense to discuss with the product team if the expectation is the expectation is that the header and search field would be reachable before the navigation panel, and create an issue for follow up.

> Aside from my one comment about the custom style changes there is one other thing i noticed with the top nav bar elements not being tabbable. Only 2 of the 4 tab options are tabbable but thats something that already exists today but we should fix before we can claim to have resolved accessibility issues

I saw this too. From my understanding this is because the `<a>` elements don't have `href` attributes and as such are not reachable. I'm not sure though as I didn't consider it part of the scope of this issue. Can we create an issue for follow up to address this?

astephanus commented on _sass/custom/custom.scss:
```diff
@@ -54,10 +54,22 @@ code {
   max-height: 100vh;
   overflow-x: hidden;
   min-width: 14rem;
+  padding-right: 1px;
+  padding-left: 1px;
+  padding-bottom: 1px;
```
@ashwin-pc I observed clipping of the keyboard focus rectangle. Adding padding seemed to resolve the issue.

astephanus commented on _sass/custom/custom.scss:
```diff
@@ -54,10 +54,22 @@ code {
+}
+
+.nav-list {
+  margin-top: 1px;
```
@ashwin-pc Same as the remarks about the padding above. I observed clipping of the keyboard focus rectangle.

ashwin-pc commented:
Yep, i'm okay with both of them being addressed in a separate issue :)

hdhalter reviewed (approved):
Thanks so much, Aaron!
